Automatic data collection for RC car + visualisation updates

alexdewar opened the pull request:
This is based on #262, so we should merge that PR before this one.

Logically this should really be two PRs, but unfortunately these two things ended up a bit tangled together (I was using the visualisation code to test the robot positioning code and fixing it up as I went along). Sorry! I hope it's not too much of a mess.

The event handling for the SFML code was messy and broken as it was: keyboard events were returned from the ``update()`` method and mouse clicks were recorded and the information could then later be retrieved. The problem is that multiple events can occur between vsyncs, so it ends up being messy trying to pick out and handle these events specifically. Instead I swapped to an interface where you can pass a lambda for event handling (which is then run inside the loop) so the caller can do whatever it wants.

Commit summary:
- Put SFML visualisation code into own namespace
- Convert remaining users of SDL-based visualiser to SFML and drop SDL support
- Fix event handling for SFML
- Tidy the interface to the ``PurePursuitController``, which we're using for data collection (no functional changes intended)
- Add ``auto_dataset_recorder`` program for automatically collecting data by driving the robot along a previously recorded route

Still to do:
- Add a simple way to collect data at a fixed distance from a previously recorded route (e.g. 1m to the left)

alexdewar committed:
SFMLWorld: Constify method

alexdewar committed:
Fix method name

alexdewar committed:
examples/simulated_ackermann: Use SFMLWorld for rendering

alexdewar committed:
examples/simulated_positioner: Allow for updating goal when in keyboard
control mode

Also remove unnecessary sleep from main loop (it'll be rate limited by
vsync anyway).

alexdewar committed:
SFMLWorld: Unify code path for updating display

alexdewar committed:
SFMLWorld: Return mouse click position in pixels

In the two places where we're using this method we actually convert back to pixels anyway.

alexdewar committed:
SFMLWorld: Use optional<> return type for mouseClickPosition()

alexdewar committed:
SFMLWorld: Use Escape cf. Q to quit

alexdewar committed:
PurePursuitController: Minor tidying

alexdewar committed:
SFMLWorld::LineStrip: allow adding vertex as sf::Vector2f

alexdewar committed:
Put SFML code into own Viz::SFML namespace

This better aligns with what we're doing with e.g. robot classes and seems tidier.

alexdewar committed:
examples/obstacle_circumnavigation: Load objects.yaml from program dir

alexdewar committed:
SFML::World: Change API to take event handler as argument

The current way of handling events is gross (and broken): mouse clicks are stored in the World object and can be retrieved later and if there's another event it's returned from the update() function. As events are handled in a loop (you have to poll through the event queue) it makes more sense to let the function take a callable thing to let the user do whatever they fancy with events.

alexdewar committed:
Add a couple of comments

alexdewar committed:
SFML: Put robot control code into own header

alexdewar committed:
SFML: Add specialised drive() method for Ackermann-type robots

alexdewar committed:
Change PurePursuitController to use optional<>s; use SFML for example

This really should have been two commits...

alexdewar committed:
Drop SDL-based code

We're not using it anywhere anymore. The SFML::World class should be used instead.

alexdewar committed:
Rename: addToWayPoint() => addWayPoint()

alexdewar committed:
examples/pure_pursuit_controller: Use FSM for managing state

alexdewar committed:
PurePositionController::getLookAheadPoint(): Remove redundant argument

Also use std::optional's emplace method

alexdewar committed:
Basic implementation of auto_dataset_recorder_simulated

It currently follows a route but doesn't save images

alexdewar committed:
Also record video

alexdewar committed:
Combine dataset_recorder and auto_dataset_recorder projects

alexdewar committed:
dataset_recorder: Put robot-specific bits into own header

alexdewar committed:
Put code for automated robot driving into own header

alexdewar committed:
Add auto_dataset_recorder

alexdewar committed:
PurePursuitController: Check added way points are valid

Also drop the setWayPoints() method as we're not using it anywhere.

alexdewar committed:
auto_dataset_recorder: Only load valid (non-NaN) coordinates from CSV file

alexdewar committed:
auto_dataset_recorder: Tidy up heading calculation code

alexdewar committed:
clang-tidy: Skip modernize-use-override check by default

It complains if we use both "virtual" and "override".

alexdewar committed:
Make arg names consistent to appease clang-tidy

alexdewar committed:
dataset_recorder: Add README

alexdewar commented:
@tenxlenx The readme file for the new dataset_recorder stuff is here: https://github.com/BrainsOnBoard/bob_robotics/blob/rc_auto_route_collection/projects/dataset_recorder/README.md

neworderofjamie commented:
Can we maybe merge #262 first, merge master in this and then review? as is, it's pretty hard to separate

alexdewar commented:
Yup, deffo. That was my intention. I just wanted to open it as a separate PR because I wanted @tenxlenx to test this code on the robot.

alexdewar committed:
Merge branch 'master' into rc_auto_route_collection

alexdewar committed:
Rename: getDistanceBetweenAxis() => getDistanceBetweenAxes()

alexdewar committed:
Specify length of RC car robot

alexdewar committed:
Fix: Start RobotRecorder's stopwatch

alexdewar committed:
Print when recording has started

alexdewar committed:
dataset_recorder: Need to explicitly allow extra CL args

alexdewar committed:
auto_dataset_recorder: Allow for aborting trial by wiggling controller

alexdewar committed:
Fix broken symlink

alexdewar committed:
RC car: Add timeout for reading RC pulses

...otherwise if the robot stops receiving pulses from the RC then it
will wait indefinitely, possibly whilst driving forwards very fast.

alexdewar committed:
Add script for plotting image databases

alexdewar committed:
Add script for setting Jetson's clock remotely

neworderofjamie commented on examples/obstacle_circumnavigation/simulated_obstacle_avoiding_positioner.cc:
```diff
@@ -80,34 +81,43 @@ int bobMain(int, char **)
 
                 // Set the goal to this position
-                const Pose2<meter_t, radian_t> goalPose{ mousePosition.x(), mousePosition.y(), 15_deg };
+                const Pose2<meter_t, radian_t> goalPose{ vec.x(), vec.y(), 15_deg };
```
is this an arbitrary angle?

alexdewar replied:
Yeah. I didn't want it to be exactly 0° because it's harder to tell if things are broken that way

neworderofjamie replied:
makes sense, might be worth a comment though

neworderofjamie commented on examples/pure_pursuit_controller/pure_pursuit_controller.cc:
```diff
@@ -16,187 +16,202 @@
-            drawLinesBetweenRects(rekt_list, display.getRenderer());
-        }
+        switch (event) {
+        case Event::Enter:
```
Ifs would probably make sense here - imo swith is only preferable when each case is a one liner otherwise it's easy to lose track of the stupid fall-through

alexdewar replied:
That's less of a problem with recent versions of gcc though. They give you a warning if you don't have ``// falls through`` or similar in there. So our CI will pick up these sorts of problems.

neworderofjamie commented on include/robots/control/pure_pursuit_controller.h:
```diff
@@ -30,44 +31,40 @@ class PurePursuitController {
     //! sets the lookahead distance. Large values causes the car to cut corners,
     //! where small values makes the car follow the path more closely
-    void setlookAheadDistance(const millimeter_t distance);
+    void setLookAheadDistance(const millimeter_t distance);
 
     //! sets the distance between the car's wheel bases
     void setWheelBaseLength(const millimeter_t length);
 
     //! sets the stopping distance. If the car is within this distance, the controller stops
     void setStoppingDistance(const millimeter_t distance);
```
some pointless ``const`` here

neworderofjamie commented on include/viz/sfml/robot_control.h:
```diff
@@ -0,0 +1,130 @@
+     * to manage state instead seems grosser.
+     */
+    static SpeedType velocity{};
+    static TurnType turn{};
```
What's the purpose of this? It seems a really bad idea as the values will be shared between multiple robots e.g. in a simulation. Why not just cache the velocity and turn values in the robot class (like tank does)? They're also useful for dead reckoning. Also, why doesn't the Ackermann robot just keep doing whatever the last command was if you don't call drive like all the other robot types?

alexdewar replied:
I actually removed the value-caching from ``TankBase`` as I felt that having the extra state in there was making things more complicated (i.e. all subclasses then had to make sure that they told ``TankBase`` when they were updating the motor commands).

The Ackermann robot *will* carry on doing whatever the last command was (``velocity`` and ``turn`` are only zero-initialised in the first call to ``drive()``). I admit that it is a bit gross to have this hidden state lurking in ``drive()`` *however* I don't really want to make an ``AckermannBase`` class which carries around this state either :shrug:. (I suppose the other option would be to make a templated class purely for the robot control.) The "multiple robots" problem is *really* unlikely to be an issue because we're never going to be controlling multiple robots with this interface anyway.

neworderofjamie commented on projects/dataset_recorder/auto_dataset_recorder.cc:
```diff
@@ -0,0 +1,174 @@
+
+        // Get initial position (try to get reading 5 times)
+        const MapCoordinate::UTMCoordinate utmStart = [&]() {
+            LOGI << "Waiting for initial GPS fix";
+            for (int i = 0; i < 5; i++) {
+                checkForErrors();
+
+                const auto data = m_GPS.read();
+                if (data && data->gpsQuality != GPS::GPSQuality::INVALID) {
+                    return MapCoordinate::latLonToUTM(data->coordinate);
+                }
+
+                std::this_thread::sleep_for(ReadDelay);
+            }
+
+            throw std::runtime_error{ "Could not get initial GPS fix" };
+        }();
```
why is this a lambda function? Just so you can make ``utmStart`` const?

alexdewar replied:
Partly. But it does also make the logic a bit clearer because we can do a ``return`` in that loop body. An alternative would be to turn this into a separate member function.

neworderofjamie commented on projects/dataset_recorder/auto_dataset_recorder_simulated.cc:
```diff
@@ -0,0 +1,91 @@
+using namespace units::math;
+namespace plt = matplotlibcpp;
+
+class RobotArrow
```
Why is this a class rather than just a function? ``plt::quiver`` is going to surely do more than enough memory allocating of its own to make any attempt to save allocations of ``m_X``, ``m_Y`` etc worthwhile. If you really care just have the vectors at ``bobMain`` scope and do this manually in the one place you actually draw an arrow!

alexdewar replied:
True true

neworderofjamie commented on projects/dataset_recorder/dataset_recorder.cc:
```diff
@@ -1,100 +1,23 @@
     CLI::App app{ "Record data with RC car robot" };
     auto *useSystemClock = app.add_flag("--use-system-clock",
                     "Use the system clock to get current date rather than GPS");
+    app.allow_extras();
```
why not just parse runtime as an optional argument?

alexdewar replied:
Good point. That would make the logic cleaner too

neworderofjamie commented on projects/dataset_recorder/robot_recorder.h:
```diff
@@ -0,0 +1,225 @@
+
+#ifdef DUMMY_CAMERA
+        // RandomInputs don't block, so throttle the framerate to 30 fps
+        std::this_thread::sleep_for(std::chrono::milliseconds(33));
```
maybe this would be a good feature to include in ``Video::RandomInput`` instead?

alexdewar replied:
Good plan

neworderofjamie commented on projects/dataset_recorder/robot_recorder.h:
```diff
@@ -0,0 +1,225 @@
+
+        // Poll for GPS data
+        if (const auto optData = m_GPS.read()) {
+            const auto &gpsData = optData.value();
```
should this be ``m_GPSData = optData.value()``? If not, I don't understand where ``m_GPSData`` is getting set

alexdewar replied:
Oooops! Good spot

neworderofjamie replied:
Yey! I actually spotted a bug!

neworderofjamie commented on projects/dataset_recorder/robot_recorder.h:
```diff
@@ -0,0 +1,225 @@
+    units::frequency::hertz_t getFrameRate() const
+    {
+#ifdef DUMMY_CAMERA
+        return units::frequency::hertz_t{ 30 };
```
Again, it would be much cleaner to just add this to ``Video::RandomInput`` (or make a derived version just used here or something) what with this already being a virtual method

alexdewar replied:
True! Done

neworderofjamie commented on projects/dataset_recorder/robot_recorder.h:
```diff
@@ -0,0 +1,225 @@
+
+        // We already have current date from GPS
+        if (currentTime.tm_year) {
+            goto out;
```
Why not just add:

```c++
std::tuple<int, int, int> getCurrentDate(bool useSystemClock)
{
     if (useSystemClock) {
        LOGW << "Using system clock to get current date. This may not be correct!";
        const auto rawTime = std::time(nullptr);
        const auto &systemTime = *gmtime(&rawTime);
        return std::make_tuple(systemTime.tm_mday, 
                               systemTime.tm_mon,
                               systemTime.tm_year);
    }
    else {
        Stopwatch sw;
        sw.start();
        while (sw.elapsed() < 5s) {
            currentTime = m_GPS.getCurrentDateTime();
            if (currentTime.tm_year) {
                return std::make_tuple(currentTime.tm_mday, 
                                       currentTime.tm_mon,
                                       currentTime.tm_year);
            }
        }

        // ...and if we didn't, give up
        throw std::runtime_error{ "Timed out waiting for date from GPS" };
    }
}
```
and then, in ``getCurrentDateTime``:
```c++
if(!currentTime.tm_year) {
    std::tie(currentTime.tm_mday, 
             currentTime.tm_mon,
             currentTime.tm_year) = getCurrentDate(useSystemClock);

}
```
You could also go a bit less modern C++ and pass ``currentTime`` by reference and set the day, month and year fields directly 🤷

alexdewar replied:
The only problem with this is that it doesn't get the timezone info. The timezone will almost certainly be UTC anyway, but we do at least parse that info...

alexdewar replied:
Actually I could just ensure that it is UTC with a ``BOB_ASSERT()`` and then do it your cleaner way....

alexdewar replied:
It occurs to me that another way around the ``goto`` ugliness is with the immediately invoked lambda idiom again :wink:

neworderofjamie replied:
your choice! I feel especially with the option of the immediately invoked lambda idiom and this being an exception-allowing codebase, there's not many reasons for a goto

neworderofjamie commented on src/robots/control/pure_pursuit_controller.cc:
```diff
@@ -1,179 +1,186 @@
+
+    std::experimental::optional<Vector2<millimeter_t>> lookAheadPoint;
+    for (unsigned int i = 0; i < m_wayPoints.size() - 1; i++) {
+        // path points
```
It would be really nice to have a little bit more explanation of this maths

alexdewar replied:
There is *some* description in the body of the loop, but tbh I just tweaked @tenxlenx's code here so I didn't write it from scratch... @tenxlenx do you have a reference to the paper you used for this that we could stick in a comment?

neworderofjamie reviewed (commented):
All minor comments aside from one bug I _think_ I spotted

alexdewar committed:
Return braced initialiser list

clang-tidy was complaining about this.

alexdewar commented on arduino/i2c_slave_code_rccar/i2c_slave_code_rccar.ino:
```diff
@@ -78,16 +78,17 @@ drive(volatile Movement &m)
+  // Timeout for reading pulses (in microsecs)
+  constexpr auto Timeout = 500000;
 
-  // curTurn is zero if controller not connected
```
No, it is (and I should probably put it back in). The 0 in this case indicates that ``pulseIn()`` timed out

alexdewar committed:
Add comment about pulseIn() timing out

alexdewar committed:
Remove pointless consts

alexdewar committed:
Video::RandomInput: Allow user to set frame rate

alexdewar committed:
examples/display: Add option to use RandomInput

alexdewar committed:
dataset_recorder: Actually set m_GPSData

alexdewar committed:
RandomInput: We don't actually need to reset the timer on each frame....

We're doing modular arithmetic here anyway so it doesn't mater.

alexdewar committed:
RandomInput: Add some setters cf. passing everything to the constructor

alexdewar committed:
Use lambda to avoid gotos

alexdewar committed:
We don't need a separate class for plotting an arrow really

neworderofjamie commented on include/video/randominput.h:
```diff
@@ -1,40 +1,56 @@
         outFrame.create(m_Size, CV_8UC3);
         fillRandom(outFrame.data, outFrame.data + m_Size.area() * 3);
+
+        waitForFrameDelay();
```
shouldn't these non-sync methods just return false if the stopwatch hasn't reached the delay time? Then the generic method will deal with the waiting for you.

alexdewar replied:
True! I wasn't thinking straight

alexdewar committed:
Add comment about non-zero angle

alexdewar committed:
RandomInput: Correctly implement async methods

alexdewar committed:
dataset_recorder: Make program run time an optional CLI argument

alexdewar committed:
Merge branch 'master' into rc_auto_route_collection

alexdewar committed:
Merge remote-tracking branch 'origin/master' into rc_auto_route_collection

alexdewar committed:
Missing override marker

alexdewar committed:
Merge branch 'master' into rc_auto_route_collection

alexdewar committed:
Merge branch 'master' into rc_auto_route_collection

alexdewar committed:
We need to explicitly invoke ImageDatabase's constructor

Presumably this worked before the merge. Not sure why it isn't working
now.

alexdewar committed:
Merge branch 'master' into rc_auto_route_collection

alexdewar committed:
Fix use of uninitialised variable

Spotted by clang-tidy.
